realm/realm-library/src/androidTest/java/io/realm/RealmMigrationTests.java

```diff
@@ -1209,6 +1211,74 @@ public void migrate(DynamicRealm realm, long oldVersion, long newVersion) {
         Realm.migrateRealm(config, migration);
     }
 
+    @Test
+    public void renameAndAddField() {
+        final String CLASS_NAME = "MigrationFieldRenameAndAdd";
+
+        RealmMigration migration = new RealmMigration() {
+            @Override
+            public void migrate(DynamicRealm realm, long oldVersion, long newVersion) {
+                realm.getSchema().get(CLASS_NAME)
+                        .renameField("string1", "string2")
+                        .addField("string1", String.class);
+            }
+        };
+
+        RealmConfiguration config = configFactory.createConfigurationBuilder()
+                .schema(MigrationFieldRenameAndAdd.class)
+                .schemaVersion(2)
+                .migration(migration)
+                .assetFile("rename-and-add.realm")
+                .build();
+
+        Realm realm = Realm.getInstance(config);
+        assertTrue(realm.getSchema().get(CLASS_NAME).hasField("string1"));
+        assertTrue(realm.getSchema().get(CLASS_NAME).hasField("string2"));
+        realm.close();
+    }
+
+    @Test
+    public void renameAndAddIndexedField() {
+        final int oldTestVal = 7;
+        final Long testVal = Long.valueOf(293);
+        final String CLASS_NAME = "MigrationIndexedFieldRenamed";
+
+        RealmMigration migration = new RealmMigration() {
+            @Override
+            public void migrate(DynamicRealm realm, long oldVersion, long newVersion) {
+                realm.getSchema().get(CLASS_NAME)
+                        .renameField("testField", "oldTestField")
+                        .addField("testField", Long.class);
+            }
+        };
+
+        RealmConfiguration config = configFactory.createConfigurationBuilder()
+                .schema(MigrationIndexedFieldRenamed.class)
+                .schemaVersion(2)
+                .migration(migration)
+                .assetFile("rename-and-add-indexed.realm")
+                .build();
+        realm = Realm.getInstance(config);
+
+        realm.beginTransaction();
+        MigrationIndexedFieldRenamed obj = realm.createObject(MigrationIndexedFieldRenamed.class, 2);
```

Comment:
Perhaps. I think the reason we originally didn't want to do that was because the rest of the dynamic API is string based, and referencing a class by class reference would break if you renamed the class during a migration since you would always reference the new name instead of the old one actually in the schema. By forcing people to use Strings they "lock" down the name to a particular point in time.

And yes I know that the above argumentation is slightly against by review comments for this PR, but hey have slightly different concerns 😄